src/main/java/net/minecraftforge/common/DimensionManager.java

```diff
@@ -412,67 +417,112 @@ public static void unloadWorlds(Hashtable<Integer, long[]> worldTickTimes)
      * block of free ids. Always call for each individual ID you wish to get.
      * @return the next free dimension ID
      */
-    public static int getNextFreeDimId() {
+    public static int getNextFreeDimId()
+    {
         int next = 0;
+        while ((next = packedIds.nextClearBit(next)) < ID_THRESHOLD)
+        {
+            if (checkAvailable(next)) return next;
+        }
         while (true)
         {
-            next = dimensionMap.nextClearBit(next);
-            if (dimensions.containsKey(next))
-            {
-                dimensionMap.set(next);
-            }
-            else
-            {
-                return next;
-            }
+            while (unpackedIds.contains(next)) next++;
+            if (checkAvailable(next)) return next;
         }
     }
 
-    public static NBTTagCompound saveDimensionDataMap()
+    private static boolean checkAvailable(int id)
     {
-        int[] data = new int[(dimensionMap.length() + Integer.SIZE - 1 )/ Integer.SIZE];
-        NBTTagCompound dimMap = new NBTTagCompound();
-        for (int i = 0; i < data.length; i++)
+        if (dimensions.containsKey(id))
+        {
+            markUsed(id);
+            return false;
+        }
+        return true;
+    }
+
+    private static void markUsed(int id)
+    {
+        if (id < ID_THRESHOLD)
+        {
+            packedIds.set(id);
```

Comment:
This will crash on negative IDs

Reply:
It's only used for non-negative IDs though - both here and in the previous code.